[[ Tests ]] Standalone test runner

montegoulding opened the pull request.

montegoulding committed:
[[ Tests ]] Expose testFilesFolder in input lib

montegoulding committed:
[[ Tests ]] Don't lcb compile in standalone runner

montegoulding committed:
[[ Tests ]] Skip securityPermsissions tests in stanadalone runner

montegoulding committed:
[[ Tests ]] Skip test requiring IDE stack in standalone runner

montegoulding committed:
[[ Tests ]] Skip lcb compile in standalone runner

montegoulding committed:
[[ Tests ]] Skip tests requiring IDE files in standalone runner

montegoulding committed:
[[ Tests ]] Delete heavily replicated test

montegoulding committed:
[[ Tests ]] Skip some click tests in standalone runner

The click command appears to fail to detect text in the standalone
test runner. This needs further investigation.

montegoulding committed:
[[ Tests ]] Use testFilesFolder for test files

montegoulding committed:
[[ Tests ]] Uniquify socket names

montegoulding committed:
[[ Tests ]] Skip docs tests in standalone runner

montegoulding committed:
[[ Tests ]] Move file to input files dir

montegoulding committed:
[[ Tests ]] Load extension only if not loaded

montegoulding committed:
[[ Tests ]] Don't load libURL in standalone runner

montegoulding committed:
[[ Tests ]] Skip test requiring repo file in standalone runner

montegoulding committed:
[[ Tests ]] Fix copy paste error

montegoulding committed:
[[ Tests ]] Fix brush tests

montegoulding committed:
[[ Tests ]] Expose GetPlatform as a public

montegoulding committed:
[[ Tests ]] Fix TestGoCardUnquotedName test

montegoulding committed:
[[ Tests ]] Fix undo delete text test

The test previously performed delete and undo on the object itself
while the test assert description indicates it is testing text deletion
undo. This patch makes it test what it's meant to test.

montegoulding committed:
[[ Emrun ]] Add --emrun flag to emscripten builds

The `--emrun` flag is only erquired when linking. It injects
some code so that if requests come from localhost it will post
stdout and stderr back to the requesting server. This will be
used by our test suite and by our HTML5 deploy library.

montegoulding commented:
@runrevmark @livecodeali This should be good to go now. I'm leaving as WIP until dependant patches have been merged and I can drop those commits from this patch.

For reference the dependant PRs are:
- #5560
- #5572

Some other patches remain in this PR until they are merged up from `develop-8.1`

montegoulding commented:
If this is merged as is we should open the following enhancement requests:

 - A method to get data back for Android & iOS (sockets?)
 - Run Android (allow specifying emulator or connected device)
 - Build iOS allow specifying provisioning profile
 - Run iOS (allow specifying simulator or use mergDeploy for connected device)
 - Revisit mergDeploy automatic run after install
 - Build & Run in server
 - Add a build target for development that builds a target platform & host platform then builds and runs the target platform standalone

montegoulding commented:
This patch required the IDE patch https://github.com/livecode/livecode-ide/pull/1635

montegoulding committed:
[[ Error ]] Improve external load fail message

montegoulding committed:
[[ Tests ]] Add TestSkipIf[Not] command

This patch adds `TestSkipIf` and `TestSkipIfNot` commands to
simplify the handling of particular features that are unupported
in a given environment. This allows us to change the condition
for skip centrally. For example, should Emscripten start supporting
wait we could turn all those tests on with a one line change.

montegoulding committed:
[[ Tests ]] Centralise common test teardown

This pathc resets the state of the environment for common
teardown tasks to what it was prior to a test at the end
of the test to reduce the need for test teardown code.

montegoulding committed:
[[ Tests ]] Improve test skip conditions

montegoulding committed:
[[ Tests ]] Test 2038 problem

montegoulding committed:
[[ Emscripten ]] Add a way to get the emscripten files

montegoulding committed:
[[ Tests ]] Add a way to set the error map

montegoulding committed:
[[ Tests ]] Add standalone test runner

This patch adds a standalone test build and run script that can be
run from command line or within the IDE. It also adds a standalone
builder test to ensure the tests can be run in sequence within the
same process without conflicting.

livecodeali commented on tests/lcs/core/engine/target.livecodescript:
```diff
@@ -58,3 +58,12 @@ end TestDeleteTargetInOwner
+   if there is a button "script" then
+      delete button "script"
+   end if
+end TestTeardown
```
Is this needed anymore?

livecodeali commented on tests/lcs/core/interface/interfaceui.livecodescript:
```diff
@@ -294,5 +348,12 @@ push current card
+   if there is a stack "Test" then
+      delete stack "Test"
+   end if
+end TestTeardown
```
None of these `delete stack`s are needed anymore are they?

montegoulding replied:
no the probably aren't

livecodeali commented on tests/lcs/core/interface/interfaceui.livecodescript:
```diff
@@ -37,16 +37,22 @@ get the formattedRect of char 1 of field "tf"
-TestAssert "test", the clickChar is "a"
+click button 3 at tPoint
+if TestIsInStandalone() then
+   TestAssertBroken "the clickChar is a", the clickChar is "a", "not sure"
```
What's the deal with this "not sure"? I think it's a bug if these aren't working in a standalone isn't it? In which case the reason should be the bug number. Or are you not sure if it is a bug or not?

montegoulding replied:
lol... yeah I really wasn't sure what was going on there. I will try testing it again to see if I can work it out.

montegoulding committed:
[[ Tests ]] Improve reset of global state

This patch improves the reset of global state after invoking a test
by moving the resset into the finally block instead of within try
and adding httpProxy, httpHeaders and stackFileVersion reset.

montegoulding committed:
[[ Tests ]] Skip canvas test if no lcb

livecodeali commented:
@livecode-vulcan review ok 99941be

livecode-vulcan commented:
💙 review by @livecodeali ok 99941beb3985fb4a7afade2ceb7f188ffa96008c

livecode-vulcan commented:
😞 test failure 99941beb3985fb4a7afade2ceb7f188ffa96008c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/535)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/550)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/524)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/467)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/470)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/530)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/538)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/533)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/550)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/533)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/543)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/565)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/529)

livecodeali commented:
@livecode-vulcan retry 99941be

livecode-vulcan commented:
😞 test failure 99941beb3985fb4a7afade2ceb7f188ffa96008c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/537)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/552)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/526)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/469)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/472)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/532)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/540)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/535)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/552)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/535)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/545)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/567)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/531)

livecodeali commented:
@livecode-vulcan retry 99941be

livecode-vulcan commented:
😎 test success 99941beb3985fb4a7afade2ceb7f188ffa96008c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/539)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/554)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/528)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/471)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/474)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/534)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/542)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/537)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/554)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/537)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/547)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/569)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/533)
